Adjust text to diagonal and vertical lines in LineZoneAnnotator and control visibility of "in" / "out"

jcruz-ferreyra opened the pull request:
### Description

In this pull request, I've made some enhancements to `supervision/detection/line_counter.py` to provide more flexibility and control over the counting behavior of the `LineZone` class and the `LineZoneAnnotator` class.

- `LineZone` Class: I've introduced a new attribute called **direction** to the `LineZone` class. This attribute allows you to specify the direction in which elements will be counted. The possible values for this parameter are:
  - **"in"**: When viewing the line horizontally with the start point to the left, the `LineZone` annotates elements crossing up the line.
  - **"out"**: Similarly, when viewing the line horizontally with the start point to the left, the `LineZone` annotates elements crossing down the line.
  - **"both"** (default value): The `LineZone` annotates both elements crossing up and down the line.
 
Importantly, please note that the **direction** attribute does not affect the behavior of the **trigger** method. The **trigger** method continues to calculate elements crossing in both directions.

The primary impact of this enhancement is on the **annotate** method within the `LineZoneAnnotator` class. By introducing this attribute to the `LineZone` class, we can customize the counting behavior for each individual `LineZone`, avoiding the need to create multiple instances of `LineZoneAnnotator`.

- `LineZoneAnnotator` class, I've introduced a new function called annotate_count that is integrated into the existing annotate method. This enhancement streamlines the process of annotating count information for elements crossing the `LineZone`. The functionality of the annotate_count function can be conceptually divided into four key steps:

  1. **Text and Image Creation**: The function takes in the text for "in" and "out" counts as input. It then generates two background square images. The count text is annotated in the center of one image, and a text box is added to the other.

  2. **Rotation to Match LineZone Orientation**: Both background images are rotated to match the orientation of the LineZone. This ensures that the annotations align accurately with the direction of the line. For this step I've used warpAffine function from cv2 package.

  3. **Positioning and Cropping**: The rotated background images are positioned within the frame, taking into account the line's position and orientation. In cases where a portion of the images extends beyond the frame, they are cropped to fit within the frame boundaries.

  4. **Direct Annotation on Frame**: The original frame is annotated with the text box and text using pixel manipulation. By directly modifying the array representing the image, the count annotations are integrated into the frame.

### Type of change

 New feature (non-breaking change which adds functionality)

### How has this change been tested, please provide a testcase or example of how you tested the change?

The following colab notebook shows the result of the changes made. In it I tried LineZones with different orientation and direction parameter (**"both"**, **"in"**, **"out"**).

https://colab.research.google.com/drive/19yobSRlqgX_s8k9_oRw0QgIipKDjs7hx?usp=sharing

### Any specific deployment considerations
No specific deployment considerations. Just one parameter added as an attribute for LineZone Class with default value "both" mantaining previous implementation.

### Docs
No changes made to Docs

SkalskiP commented:
Hi, @jcruz-ferreyra 👋🏻! Still want to help us out?

jcruz-ferreyra commented:
Hey @SkalskiP, I would be glad to contribute to the project!

SkalskiP commented:
Awesome! If you are still around, I'll do my best to make a PR review today.

jcruz-ferreyra commented:
Excellent! I will be waiting for it.

SkalskiP commented:
@jcruz-ferreyra, is there any chance you could update your project fork so that your `develop` would be in sync with `supervision/develop` and then merge `develop` into your feature branch? It looks like `pre-commit.ci ` won't run as you created this fork quite far ago.

jcruz-ferreyra commented:
Okay! Should I still make pull requests to the `redesigned_annotators` branch?

SkalskiP commented:
@jcruz-ferreyra I just updated the target branch to `develop`, but it looks like `pre-commit.ci` is still not there.

SkalskiP commented:
@jcruz-ferreyra I hope this PR is closed only for technical reasons. 🤞🏻

jcruz-ferreyra commented:
Yes, I am merging and reopening it in a few minutes!

jcruz-ferreyra commented:
@SkalskiP Done!

SkalskiP commented on supervision/detection/line_counter.py:
```diff
@@ -13,19 +14,23 @@ class LineZone:
     """
 
-    def __init__(self, start: Point, end: Point):
+    def __init__(self, start: Point, end: Point, direction="both"):
```
I love your added `direction` argument. I don't like `direction` as a string. We can solve that by:

1. Using enum

    ```python
    class Direction(Enum):
        IN = "in"
        OUT = "out"
        BOTH = "both"
    ```
    
    than:
    
    ```python
    def __init__(self, start: Point, end: Point, direction: Direction):
    ```

2. Using two separate bool arguments (preferred):

    ```python
    def __init__(
        self, 
        start: Point, 
        end: Point, 
        trigger_in: bool = True, 
        trigger_out: bool = True
    ):
    ```

Please update the docstring as well to reflect that change.

SkalskiP commented on supervision/detection/line_counter.py:
```diff
@@ -113,13 +118,200 @@ def annotate(self, frame: np.ndarray, line_counter: LineZone) -> np.ndarray:
 
         """
+
+        def annotate_count(text, text_over=True):
```
Let's not define that function inside another function. Let's convert this into a private `_annotate_count` method in the `LineZoneAnnotator` class.

SkalskiP replied:
Let's also add argument types.

SkalskiP commented:
Hi @jcruz-ferreyra 👋🏻, I'm super excited about this change. Left a few comments in the code. Is it possible to divide the code in `annotate_count` into smaller functions? It is tough to follow. That function alone is 200 lines. It would be a lot easier to understand what is going on if you could extract some parts of the logic into separate functions and just call them in `annotate_count`.

SkalskiP commented:
Hi @jcruz-ferreyra 🤞🏻 understood! Can’t wait to have it in develop!

SkalskiP commented:
Hi @jcruz-ferreyra, It looks like we won't be able to merge it before release tomorrow?

jcruz-ferreyra commented:
Hi @SkalskiP! I finished with the code and am currently documenting the new methods. I am commiting asap!

jcruz-ferreyra commented:
Hi @SkalskiP! I recently pushed the requested changes. Let me know if the code is straightforward enough or more comments are needed. Functionality can be tested in the colab notebook of the original post.

SkalskiP commented:
Hi, @jcruz-ferreyra 👋🏻 Thanks for that update! I took a quick look, and it already looks a lot better. I'll try to take a deeper look on Monday. 🙏🏻

jcruz-ferreyra commented:
Hi @SkalskiP! Are you still interested in these changes? Best regards!

SkalskiP commented:
Hi, @jcruz-ferreyra 👋🏻 We recently made many updates in `LineZone` taking the counting logic to the next level. And yes, I still want this PR to be merged, but I felt a bit stupid to ask you one more time to make it compatible so I was thinking that I'll just update it myself not to bother you...

jcruz-ferreyra commented:
Don´t worry @SkalskiP, I can do it! Will push it again in these days.

SkalskiP commented:
Awesome! 🔥 Thank you, and sorry. We are putting a lot of effort into making LineZones better lately. We plan to expose it more in docs.

SkalskiP commented:
@jcruz-ferreyra I hope you just decided to close for now and reopen once you will make it work with updated API?

jcruz-ferreyra commented:
> @jcruz-ferreyra I hope you just decided to close for now and reopen once you will make it work with updated API?

Yes, of course!

jcruz-ferreyra commented:
Hi @SkalskiP, I've updated my branch to incorporate the latest API adjustments.

In the `_annotate_count` method, I now create an auxiliary image for count annotation. I then rotate it to align with the `LineZone` object, adding it to the frame with an alpha channel for opacity (emulating PNG functionality).

Notably, I've switched to using the `display_in_count: bool` and `display_out_count: bool` attributes, introduced in the updated API. These replace the earlier `trigger_in` and `trigger_out` attributes discussed previously in this conversation.

Additionally, I've introduced two new attributes in the LineZoneAnnotator class: draw_text_box: bool and draw_centered: bool. You can test their functionality in the colab notebook provided in the initial pull request message, available [here](https://colab.research.google.com/drive/19yobSRlqgX_s8k9_oRw0QgIipKDjs7hx?usp=sharing).

Feel free to review, and I welcome any feedback.

onuralpszr commented:
@jcruz-ferreyra  I know you doing changes but we already include in/out visibility and you are re-introduce this plus when I read codes it makes my lifes hard plus you have 29 commits I mean it is okay but I don't want to have messy history as well. Please understand me I am not trying to make your life harder but I was expecting more minimal change and maybe 2-3 commits not 29 and this huge change slice

jcruz-ferreyra commented:
Hello @onuralpszr,

I appreciate your feedback and I understand the concern about the number of commits and the code changes.

I noticed that with each update to this pull request, a new merging commit was created, and the commit history became even more extensive due to format issue resolutions. To address this, I'm considering cleaning up the pull request by removing unnecessary commits, keeping only those that matter. How does that sound to you?

Regarding the changes included, I find it a bit challenging to reduce the number of them. The six private methods in the PR are essential for achieving aligned text with cv2, especially since cv2.putText() doesn't support angled text natively. However, I'll try my best to minimize trivial changes to ensure the PR is as clean as possible. Would that be OK?

Thanks again for your feedback!

SkalskiP commented:
Hi @jcruz-ferreyra 👋🏻 

> To address this, I'm considering cleaning up the pull request by removing unnecessary commits, keeping only those that matter (May be start a fresh PR from a new branch).

Opening a new PR will make the most sense.

>  I'll try my best to minimize trivial changes to ensure the PR is as clean as possible.

Thank you. We will use the review process to make those changes as efficient as possible.

Thanks @onuralpszr and @jcruz-ferreyra 🙏🏻

jcruz-ferreyra commented:
Hi @SkalskiP, @onuralpszr. I've created a new [Pull request](https://github.com/roboflow/supervision/pull/854). I will close this one.

SkalskiP commented:
@jcruz-ferreyra, thanks a lot 🙏🏻 ! I'll try to make the first round of reviews as soon as possible.
